module4-authentication-and-security/r1.1.1-authentication-persistence/README.md

```diff
@@ -0,0 +1,268 @@
+
+![JWT Diagram](../assets/use-of-jwt.webp)
+
+### **JWT Validation**
+
+A middleware library like [express-jwt](https://www.npmjs.com/package/express-jwt) can be used on the server to validate JWTs.
+
+```js
+// ./app.js
+
+const express = require("express");
+const jwt = require("express-jwt");
+
+// Init our express app
+const app = express();
+
+// Below is where we introduce JWT
+// Make sure your app secret is unique and
+// defined by cryptographic random generator.
+app.use(
+  jwt({
+    secret: "MY_APP_SECRET",
+    algorithms: ["HS256"],
+  })
+);
+
+let listener = app.listen(5000, function () {
+  console.log("Listening on port " + listener.address().port);
+});
+```
+
+With this, we've set up a JWT validator for every request, that reads the `Authorization` header, validates the token against the signature and expiration date, scope, etc, and parses the claims to identify the user.
+
+Then e can simply access the user object in the request as follows:
+
+```js
+const user = req.user;
+```
+
+### **Issuing JWT**
+
+To issue JWT, your authentication server can use a library [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken) to issue the token as follows:
+
+```js
+// ./routes/auth.js
+...
+const jwt = require('jsonwebtoken');
+
+// Signing secret
+const SECRET = 'MY_APP_SECRET';
+
+
+router.post('/login', async (req, res) => {
+  const {username, password} = req.body;
+
+  const user = await User.login(username, password);
+
+  if(user) {
+    // User is valid, issue JWT
+    const claims = {
+      sub: user.id,
+      name: user.name,
+      username: user.username
+    }
+
+    const token = jwt.sign(claims, SECRET, { expiresIn: "2h" });
+
+    res.json({ token })
+  } else {
+    res.status(400).render('login')
+  }
+})
+
+...
+```
+
+> ⚠️ **Warning**: Please note that in the code above, for clarity, the `SECRET` was hard-coded and defined inline. **Never do this**. Usually, your `app-secret` is stored in environment variables and **never** shipped with the code.
+
+> 💡 **TIP**: Always keep your tokens short-lived as they are, by design, harder to revoke. The token by default will stay functional until it expires.
+
+### **JWT Signature**
+
+You're probably wondering: if JWTs can be easily decoded and their payload can be seen, how do we make sure that someone doesn't pass any payload to authenticate a request? The signature is the answer.
+
+When issuing JWTs, we take the claims and encrypt them using the `app-secret` key. Then attach that signature to the JWT as a `base64` encoded string. Remember that the `app-secret` is only known by your app.
+
+When the JWT is passed back to the server in the header, we check the claims, and sign them again using our `app-secret`, then compare this signature with the signature provided by the token. If they match, that means the JWT is indeed issued by this app. Otherwise, it is fake.
```

Comment:
this part needs to be explained better, i would explain it from the perspective of someone trying to give a token that was encrypted by another app secret (not yours). since there has been no explanation of basic encryption in any previous sections, again line 181 to 183 will be mega confusing to a student

Reply:
I don't understand this. I wouldn't dive into how encryption works or the math behind it, and for now offer ways to encrypt only. Encryption and validation are both done above, and in case the signature is incorrect, `express-jwt` would return invalid token error to be handled.

If I didn't understand correctly please elaborate.